components/script/dom/vrdisplay.rs

```diff
@@ -909,6 +935,53 @@ impl VRDisplay {
             pair.1 = None;
         }
     }
+
+    /// Initialize XRInputSources
+    fn initialize_inputs(&self) {
+        if self.initialized_inputs.get() {
+            return;
+        }
+        self.initialized_inputs.set(true);
+
+        let (sender, receiver) = ipc::channel(self.global().time_profiler_chan().clone()).unwrap();
+        let display = self.display.borrow().display_id;
+        self.webvr_thread()
+            .send(WebVRMsg::GetGamepadsForDisplay(display, sender))
+            .unwrap();
+        match receiver.recv().unwrap() {
+            Ok(gamepads) => {
+                let global = self.global();
+                let session = self
+                    .xr_session
+                    .get()
+                    .expect("initialize_inputs called on a VR session");
+                let roots: Vec<_> = gamepads
+                    .into_iter()
+                    .map(|g| {
+                        (
+                            g.1.gamepad_id,
+                            XRInputSource::new(&global, &session, g.0, g.1),
+                        )
+                    })
+                    .collect();
+
+                let mut inputs = self.input_sources.borrow_mut();
+                for (id, root) in &roots {
```

Comment:
Nit: can we avoid collecting the inputs into a vec and then discarding it?

Reply:
I think there were some rooting considerations here, previously, but they may have been cleaned up.